### PR TITLE
feat: support compression on gRPC server

### PR DESCRIPTION
**Pull request by shuiyisong**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This pr mainly supports gRPC server compression encoding options(`gzip`, `zstd`). Now gRPC server is allowed to receive requests with such encoding.

In this pr we also enable zstd compression encoding by default between frontend and datanode.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/common/grpc/src/channel_manager.rs [hunk 1]
```diff
@@ -261,6 +262,8 @@ pub struct ChannelConfig {
     pub max_recv_message_size: ReadableSize,
     // Max gRPC sending(encoding) message size
     pub max_send_message_size: ReadableSize,
+    // Enable gzip compression for gRPC
```
Comment (fengjiachun) [inline src/common/grpc/src/channel_manager.rs:265]: Maybe:
```suggestion
    // Supported compression encoding for gRPC, e.g: gzip, zstd.
```
If I understand this code correctly.

File: src/servers/src/grpc.rs [hunk 2]
```diff
@@ -64,13 +66,33 @@ pub struct GrpcServerConfig {
...
+    // Enable gzip compression for gRPC
+    pub accept_compressed: Vec<CompressionEncoding>,
+}
+
+pub fn parse_grpc_compression_encoding(
+    encodings: &Vec<String>,
+) -> Result<Vec<CompressionEncoding>> {
+    let mut result = Vec::with_capacity(encodings.len());
+    for encoding in encodings {
+        let encoding = match encoding.as_str() {
```
Comment (fengjiachun) [inline src/servers/src/grpc.rs:78]: Is it necessary to convert `encoding` to lowercase before matching?

**Review by fengjiachun — APPROVED** [review]

File: src/servers/src/grpc.rs [hunk 3]
```diff
@@ -64,13 +66,33 @@ pub struct GrpcServerConfig {
...
+            "gzip" => CompressionEncoding::Gzip,
+            "zstd" => CompressionEncoding::Zstd,
+            _ => {
+                return UnsupportedGrpcCompressionEncodingSnafu { encoding }.fail();
```
Comment (waynexia) [inline src/servers/src/grpc.rs:82]: Maybe we should maintains a table for all supported methods in docs

File: src/servers/src/grpc/builder.rs [hunk 4]
```diff
@@ -41,15 +41,21 @@ use crate::query_handler::OpenTelemetryProtocolHandlerRef;
...
-                .max_encoding_message_size(max_send_message_size),
-        )
+        let mut service_builder = $service
+            .max_decoding_message_size(max_recv_message_size)
+            .max_encoding_message_size(max_send_message_size);
+
+        for encoding in $accept_compressed {
+            service_builder = service_builder
+                .accept_compressed(*encoding)
+                .send_compressed(*encoding);
```
Comment (waynexia) [inline src/servers/src/grpc/builder.rs:55]: Should this behavior use another config entry?

**Review by waynexia — COMMENTED** [review]
🚀

File: config/datanode.example.toml [hunk 5]
```diff
@@ -29,6 +29,9 @@ rpc_max_recv_message_size = "512MB"
 ## The maximum send message size for gRPC server.
 rpc_max_send_message_size = "512MB"
 
+## Supported compression encoding for gRPC server. Currently, `gzip` and `zstd` are supported.
+rpc_accept_compressed = ["gzip"]
```
Comment (sunng87) [inline config/datanode.example.toml:33]: as discussed in slack, I think it's not necessary to configure this, especially for accepting client data

File: src/common/grpc/src/channel_manager.rs [hunk 6]
```diff
@@ -381,6 +385,16 @@ impl ChannelConfig {
         self.client_tls = Some(client_tls_option);
         self
     }
+
+    /// Enable compression for gRPC
+    ///
+    /// Disabled by default.
+    pub fn accept_compressed(self, compression: Vec<CompressionEncoding>) -> Self {
```
Comment (tisonkun) [inline src/common/grpc/src/channel_manager.rs:392]: I don't find the logic a config option can manipulate this option.

Also, is it possible that we enable compression by default? It seems there is no downside for doing so.

Comment (shuiyisong) [reply src/common/grpc/src/channel_manager.rs:392]: It's confirmed that `channel_manager` and `client` is only used internally, we can remove those configs now. Should we also enable gzip/zstd compression encoding by default internally(between frontend/datanode/metasrv)? @sunng87 @waynexia @fengjiachun 

> Also, is it possible that we enable compression by default? It seems there is no downside for doing so.

Now the compression support is auto-enabled on our server(namely frontend), but it depends on the client request to determine whether to use compression or not(the compression encoding is done on client side).

Comment (tisonkun) [reply src/common/grpc/src/channel_manager.rs]: > Now the compression support is auto-enabled on our server(namely frontend), but it depends on the client request to determine whether to use compression or not(the compression encoding is done on client side).

Of course. We can "allow" the support. Whether or not it's compressed is based on the request - this is normal :D

**Review by tisonkun — APPROVED** [review]

**Review by waynexia — APPROVED** [review]

**Pull request merged**